### PR TITLE
Extract metadata from CITATION.cff

### LINKED ISSUE #89: Metadata extraction from CITATION.cff file
At the moment, metadata are extracted "only" from the dataset_description.json files. However, metadata can be extracted from the CITATION.cff file, which is currently not being processed.

To comply with the [BIDS validation rules](https://github.com/bids-standard/bids-specification/blob/master/src/schema/rules/checks/dataset.yaml), metadata should be extracted from CITATION.cff with priority when such a file is present.

@Peyman-N Feel free to assign this issue to me if you don’t have time to handle it.

### PR DESCRIPTION
**Pull request by Raphael-Gazzotti**
Extracts metadata from CITATION.cff including:
- DOI
- authors (with ORCID, email, affiliation)
- license
- title -fullName, shortName-
- versionIdentifier

Also lays groundwork for extracting additional information such as preferred-citation.

Closes #89

### CODE REVIEW

**Comment by Peyman-N** [comment]
The pull request seems reasonable. Can you add some tests for it, please?

File: bids2openminds/main.py [hunk 1]
```diff
@@ -184,21 +201,36 @@ def create_openminds_age(data_subject):
...
+    digital_identifier = None
+    name = None
+    version_identifier = None
+    if citation:
+        if 'doi' in citation:
+            digital_identifier = omcore.DOI(identifier=citation['doi'])
+        if 'license' in citation:
+            for lic in omcore.License.instances():
+                if citation['license'] == getattr(lic, "short_name", None):
+                    license = lic
```
[the remark this answers is not in the document]
Comment (apdavison) [reply bids2openminds/main.py]: this breaks out of the loop once we've found a match

**Review by apdavison — CHANGES_REQUESTED** [review]

File: bids2openminds/main.py [hunk 2]
```diff
@@ -48,8 +48,25 @@ def create_openminds_person(full_name):
...
+    if "authors" in dataset_description:
+        person_list = dataset_description["authors"]
+        openminds_list=[]
+        for person in person_list:
+            person_orcid = None
+            person_affiliation = None
+            if 'orcid' in person:
+                person_orcid = omcore.ORCID(identifier=person['orcid'])
+            if 'affiliation' in person:
+                person_affiliation = omcore.Affiliation(person['affiliation'])
```
[the remark this answers is not in the document]
Comment (Raphael-Gazzotti) [reply bids2openminds/main.py]: Indeed, I’ve updated it. However, there’s an implementation choice to make: should we treat everything in the affiliation field as a single, unique organization, or should we implement rules to distinguish between them?

For now, I’ve gone with the first option, considering everything in affiliation as a single organization, but it’s probably safer to not consider the affiliation field at the moment.

**Review by Peyman-N — APPROVED** [review]

**Review by lzehl — APPROVED** [review]
@apdavison or @Peyman-N please one of you double check. It looks good from my side and can be merged by the next one confirming the PR

File: bids2openminds/converter.py [hunk 3]
```diff
@@ -27,10 +28,11 @@ def convert(input_path,  save_output=False, output_path=None, multiple_files=Fal
 
     subjects_id = bids_layout.get_subjects()
 
-    # imprting the dataset description file containing some of the
+    # importing the dataset description file containing some of the metadata
     dataset_description_path = utility.table_filter(layout_df, "description")
-
     dataset_description = utility.read_json(dataset_description_path.iat[0, 0])
+    citation_path = utility.table_filter(layout_df, "CITATION")
```
[the remark this answers is not in the document]
Comment (Raphael-Gazzotti) [reply bids2openminds/converter.py]: It should return an empty DataFrame, the line just after is checking if the DataFrame is empty.

File: bids2openminds/converter.py [hunk 4]
```diff
@@ -18,6 +19,8 @@
     ("participants.tsv", "participants", ".tsv"),
     ("participants.json", "participants", ".json"),
     ("CHANGES", None, None),
+    ("CITATION", "CITATION", None),
```
Comment (apdavison) [inline bids2openminds/converter.py]: does this mean if a file named CITATION exists (which may just contain plain text) it will be handled as if it was a CITATION.cff file?

File: bids2openminds/converter.py [hunk 5]
```diff
@@ -77,10 +80,11 @@ def convert(input_path,  save_output=False, output_path=None, multiple_files=Fal
 
     subjects_id = bids_layout.get_subjects()
 
-    # imprting the dataset description file containing some of the
+    # importing the dataset description file containing some of the metadata
     dataset_description_path = utility.table_filter(layout_df, "description")
-
     dataset_description = utility.read_json(dataset_description_path.iat[0, 0])
+    citation_path = utility.table_filter(layout_df, "CITATION")
+    citation = yaml.safe_load(open(citation_path.iat[0, 0], encoding="utf-8")) if not citation_path.empty else None
```
Comment (apdavison) [inline bids2openminds/converter.py]: The file is opened but never explicitly closed. Better to use a context manager.

```suggestion
    citation = None
    if not citation_path.empty:
        with open(citation_path.iat[0, 0], encoding="utf-8") as fp:
            citation = yaml.safe_load(fp) 
```

File: bids2openminds/main.py [hunk 6]
```diff
@@ -185,21 +212,34 @@ def create_openminds_age(data_subject):
...
+    license = None
+    digital_identifier = None
+    name = None
+    version_identifier = None
+    if citation:
+        if 'doi' in citation:
+            digital_identifier = omcore.DOI(identifier=citation['doi'])
+        if 'license' in citation:
+            license = omcore.License.by_name(citation['license'])
+        if 'title' in citation:
```
Comment (apdavison) [inline bids2openminds/main.py:235]: if "title" is not present, `name` will stay as None, but probably you should fall back to taking it from `dataset_description["Name"]`

File: bids2openminds/main.py [hunk 7]
```diff
@@ -185,21 +212,34 @@ def create_openminds_age(data_subject):
...
     # dataset_type=bids2openminds_instance(dataset_description.get("DatasetType",None))
+    license = None
+    digital_identifier = None
+    name = None
+    version_identifier = None
+    if citation:
+        if 'doi' in citation:
+            digital_identifier = omcore.DOI(identifier=citation['doi'])
+        if 'license' in citation:
+            license = omcore.License.by_name(citation['license'])
```
Comment (apdavison) [inline bids2openminds/main.py:230]: What happens if citation['license'] doesn't match any openMINDS License instance?

File: pyproject.toml [hunk 8]
```diff
@@ -6,7 +6,8 @@ dependencies = [
   "openminds >= 0.2.3",
   "click>=8.1",
   "pandas",
-  "nameparser >= 1.1.3"
+  "nameparser >= 1.1.3",
+  "PyYAML"
```
Comment (apdavison) [inline pyproject.toml]: should probably pin this to a minimal version, e.g., "PyYAML >= 6.0"

File: test/test_dataset_version.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
+import pytest
+
+import pandas as pd
+from openminds import Collection
+import openminds.v4.core as omcore
+
+from bids2openminds.main import create_dataset_version
+
+def test_create_dataset_version_citation():
```
Comment (apdavison) [inline test/test_dataset_version.py:17]: please also add a test for `citation=None`, to ensure the alternative path still works.

**Review by apdavison — CHANGES_REQUESTED** [review]
I noticed a few more things.

File: bids2openminds/converter.py (hunk 4 above)
Comment (Raphael-Gazzotti) [reply bids2openminds/converter.py]: I double-checked, and you are right, apparently only CITATION.cff should be used. I will remove this one.

**Review by apdavison — APPROVED** [review]
